app/models/sample_form.rb

```diff
@@ -87,15 +87,22 @@ def save
 
     # validate forms. stop if invalid
     form_valid = self.valid?
+    self.assay_attachments.each do |aa|
+      form_valid &= aa.valid?
+    end 
+    self.notes.each do |n|
+      form_valid &= n.valid?
+    end 
```

Comment:
Hum, ActiveRecord has `validates_associated` and Sample will validate these two associations using it, but here we only have an ActiveModel so we don't have this helper. Yet, we can use [validates_each](https://api.rubyonrails.org/v4.2/classes/ActiveModel/Validations/ClassMethods.html#method-i-validates_each), so that `valid?` would always tell us whether the form is valid or not:

```ruby
class SampleForm
  validates_each :assay_attachments, :notes do |record, attr, value|
    errors.add(attr, "are invalid") unless value.all?(&:valid?)
  end
end
```